File: Data-Structure-I/36-Valid-Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        for(int i=0;i<9;++i){
            vector<int> row(10,0), col(10,0), sub_box(10,0);
            for(int j=0;j<9;++j){
                
                if(board[i][j] != '.'){
                    int value = board[i][j] - '0';
                    row[value]++;
                    if(row[value]>1) return false;
                }
                
                if(board[j][i]!='.'){
                    int value = board[j][i] - '0';
                    col[value]++;
                    if(col[value]>1) return false;
                }
                
                int x = 3*(i/3) + j/3, y = 3*(i%3) + j%3;
                if(board[x][y]!= '.'){
                    int value = board[x][y] - '0';
                    sub_box[value]++;
                    if(sub_box[value]>1) return false;
                }
            }
        }
        return true;
    }
};

File: Data-Structure-I/36-Valid-Sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = collections.defaultdict(set)
        col = collections.defaultdict(set)
        sub_box = collections.defaultdict(set)
        
        for i in range(9):
            for j in range(9):
                value = board[i][j]
                if value == '.':
                    continue
                if (value in row[i]) or (value in col[j]) or (value in sub_box[(i//3, j//3)]):
                    return False
                
                row[i].add(value)
                col[j].add(value)
                sub_box[(i//3, j//3)].add(value)
        return True
          
